Feat[#34]: 온보딩 페이지 최종 제출 시 사용자 벡터 저장

likerhythm opened the pull request:
## 확인 사항
- [ ] 💯 테스트는 잘 통과했나요?
- [x] 🏗️ 빌드는 성공했나요?
- [x] 🧹 불필요한 코드는 제거했나요?
- [x] 💭 이슈는 등록했나요?
- [x] 🏷️ 라벨은 등록했나요?

## 작업 내용
- 온보딩 최종 제출 시 사용자 정보를 저장하는 기능을 구현했습니다.
- FastAPI 서버와 통신한 후 사용자의 vector 값을 저장합니다.
- 최종적으로 사용자의 name, age, gender, weight, vector가 저장됩니다.

## 시퀀스 다이어 그램
![image](https://github.com/user-attachments/assets/48844aaa-e1d5-479e-b857-90188428bbfd)


## 주의사항
- DB 세팅 후 테스트 해 볼 예정입니다. 아직 제대로 동작하지 않을 수 있습니다.
- UserService에 vector와 weight 등 관련 없는 로직이 많아서 책임 분리가 필요하다고 생각합니다. vector와 weight 패키지를 따로 마련하는 방향으로 리팩토링 목록에 추가해 놓겠습니다.

Closes #{34}

likerhythm committed:
feat[#34]: 에러코드 추가

likerhythm committed:
chore[#34]: User 엔티티 Setter 추가

likerhythm committed:
feat[#34]: 온보딩 제출 request dto 추가

likerhythm committed:
feat[#34]: UserService 추가

likerhythm committed:
feat[#34]: UserController 추가

likerhythm committed:
feat[#34]: 온보딩 제출 결과 dto 추가

likerhythm committed:
refactor[#34]: OnboardingResponseDto로 변경

likerhythm committed:
feat[#34]: PreferMetaInfoRepository 추가

likerhythm committed:
feat[#45]: FastApi 서버로부터 온보딩 결과 받는 dto 추가

likerhythm committed:
refactor[#45]: FastApiOnboardingResponseDto 적용

likerhythm committed:
feat[#45]: UserErrorCode 추가

likerhythm committed:
feat[#45]: 온보딩 결과 받는 dto 추가

likerhythm committed:
feat[#45]: PreferMetaInfoRepository 추가

likerhythm committed:
reafctor[#45]: User 엔티티 setter 추가

likerhythm committed:
feat[#45]: UserService 추가

사용자의 age, gender, name 등 기본 사항 저장
사용자의 vector, weight 등 vector 관련 사항 저장

likerhythm committed:
Merge branch 'fix/integerate-crud-to-spring' into feat/onboarding-submit

likerhythm committed:
refactor[#34]: 가중치 전략을 spring 서버에 두는 것으로 수정

likerhythm committed:
Merge branch 'develop' into feat/onboarding-submit

likerhythm committed:
feat[#34]: MetaInfoRepository 추가

likerhythm committed:
feat[#34]: 컨텐츠 아이디로 장르 조회하는 쿼리 추가

likerhythm committed:
refactor[#34]: 가중치 전략 패턴을 WeightManager로 대체

likerhythm committed:
chore[#34]: User getter 추가

likerhythm committed:
refactor[#34]: 가중치 타입 double로 변경

likerhythm committed:
feat[#34]: 사용자 가중치 및 벡터 저장

likerhythm committed:
feat[#34]: MetaInfoErrorCode 추가

likerhythm committed:
chore[#34]: 불필요 줄 삭제

likerhythm committed:
chore[#34]: 년도를 나이로 변환하는 로직 추가

likerhythm committed:
refactor[#34]: 사용하지 않는 dto 삭제

likerhythm committed:
chore[#34]: 메서드 설명 수정

likerhythm committed:
chore[#34]: 간격 수정

sangyunpark99 commented on src/main/java/org/highfive/backend/content/entity/repository/MetaInfoRepository.java:
```diff
@@ -0,0 +1,17 @@
+    SELECT m
+    FROM MetaInfo m
+    WHERE m.name = :name AND m.type = 'GENRE'
+""")
```
@Query("SELECT m FROM MetaInfo m WHERE m.name = : name AND m.type = 'GENRE')
이정도 쿼리 길이는 한문장으로 해도 괜찮을거 같습니다. 😳

likerhythm replied:
한 줄로 하니까 깔끔해서 보기 좋네요! 반영했습니다.

sangyunpark99 commented on src/main/java/org/highfive/backend/content/exception/MetaInfoErrorCode.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public int getCode() {
+        return 0;
```
```java
@Override
    public int getCode() {
        return this.code;
    }
```
로 수정해야 나중에 예외 상황일때, 사용할 수 있습니다.

likerhythm replied:
오버라이드만 하고 빠뜨렸네요. 반영했습니다!

sangyunpark99 commented on src/main/java/org/highfive/backend/content/exception/MetaInfoErrorCode.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public String getMessage() {
+        return "";
```
this.message

sangyunpark99 commented on src/main/java/org/highfive/backend/content/exception/MetaInfoErrorCode.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public HttpStatus getHttpStatus() {
+        return null;
```
this.httpStatus

sangyunpark99 commented on src/main/java/org/highfive/backend/global/client/fastapi/FastApiClient.java:
```diff
@@ -26,17 +27,17 @@ public class FastApiClient {
+        HttpEntity<List<Long>> request = new HttpEntity<>(contentIds, headers);
 
-        ResponseEntity<Boolean> response = restTemplate.postForEntity(url, request, Boolean.class);
+        ResponseEntity<FastApiOnboardingResponseDto> response = restTemplate.postForEntity(url, request, FastApiOnboardingResponseDto.class);
```
네트워크 오류로 인해서 응답이 안오는 경우가 있을 수 있으니, 해당 부분에서 발생하는 예외를 처리해주세요 🌞

likerhythm replied:
네트워크 IO와 데이터베이스 IO는 예외 처리 신경쓰겠습니다.

likerhythm replied:
반영했습니다!

sangyunpark99 commented on src/main/java/org/highfive/backend/user/UserController.java:
```diff
@@ -0,0 +1,24 @@
+    private final UserService userService;
+
+    @PatchMapping("/userInfo")
+    public Response<Void> submitOnboarding(@Valid @RequestBody SubmitOnboardingRequestDto request) {
```
깃 컨벤션에 따라 파라미터에 final 키워드를 붙여주세요 😎

likerhythm replied:
세세한 리뷰 감사합니다!

sangyunpark99 commented on src/main/java/org/highfive/backend/user/UserController.java:
```diff
@@ -0,0 +1,24 @@
+    @PatchMapping("/userInfo")
+    public Response<Void> submitOnboarding(@Valid @RequestBody SubmitOnboardingRequestDto request) {
+        userService.initUser(request);
+        return new Response<>(20000, null, null);
```
SuccessCode enum을 활용해서 작성해주세요 🧐

likerhythm replied:
넵 반영 완료했습니다!

sangyunpark99 commented on src/main/java/org/highfive/backend/user/entity/User.java:
```diff
@@ -17,6 +19,8 @@
 import java.util.List;
 
 @Entity
+@Setter
```
@Setter를 사용하지 말고 @DynamicUpdate와 업데이트할 컬럼이 포함된 메서드를 따로 만들어서 사용하는게 좋을것 같습니다.

코드 예시
```java
public void updateBasicInfo(String name, int age, Gender gender) {
      this.name = name;
      this.age = age;
      this.gender = gender;
}
```

likerhythm replied:
좋은 방법이네요 인사이트 얻어 갑니다~!

sangyunpark99 commented on src/main/java/org/highfive/backend/user/UserService.java:
```diff
@@ -0,0 +1,98 @@
+        int age = Year.now().getValue() - request.year();
+        Gender gender = request.gender();
+        String name = request.name();
+        user.setAge(age);
```
setter 대신 update 메서드를 사용하면, 코드도 한줄로 줄어드는 이점이 있습니다.

sangyunpark99 commented on src/main/java/org/highfive/backend/user/UserService.java:
```diff
@@ -0,0 +1,98 @@
+
+        // 벡터 저장
+        String vector = response.vector();
+        user.setEmbedding(vector);
```
update 메서드를 따로 선언해서 사용하면 좋을 것 같습니다.

sangyunpark99 commented on src/main/java/org/highfive/backend/user/UserService.java:
```diff
@@ -0,0 +1,98 @@
+
+    private Map<String, Integer> countGenre(final List<Long> contentIds) {
+        final Map<String, Integer> genreCount = new HashMap<>();
+        for (long contentId : contentIds) {
```
한번의 쿼리로 해당 부분을 구현할 수 있습니다.
```java
@Query("SELECT new map(c.id as contentId, m.name as genreName) " +
       "FROM MetaInfoContents mic " +
       "JOIN mic.metaInfo m " +
       "JOIN mic.content c " +
       "WHERE m.type = 'GENRE' AND c.id IN :contentIds")
List<Map<String, Object>> findContentGenresByContentIds(@Param("contentIds") List<Long> contentIds);


Map<String, Integer> genreCount = new HashMap<>();

List<Map<String, Object>> results = contentRepository.findContentGenresByContentIds(contentIds);

for (Map<String, Object> row : results) {
    String genreName = (String) row.get("genreName");
    genreCount.put(genreName, genreCount.getOrDefault(genreName, 0) + 1);
}
```

sangyunpark99 reviewed (commented):
아쉽지만 수정해주세요

sangyunpark99 reviewed (changes requested):
리뷰한 부분 수정 부탁드려요~

likerhythm committed:
chore[#34]: 쿼리 깔끔하게 수정

likerhythm committed:
chore[#34]: getter 수정

likerhythm committed:
refactor[#34]: 예외 처리 추가 및 전용 메서드 추가

아래 FastApiCode를 추가했습니다.
- FAST_API_RESPONSE_NULL
- FAST_API_RESPONSE_ERROR
- FAST_API_CONNECTION_ERROR
- FAST_API_ERROR

또한 executeWithFastApiHandling 메서드를 두어서 예외 처리 코드를 하나의 메서드의 책임으로 두었습니다.

likerhythm committed:
chore[#34]: final 키워드 추가

likerhythm committed:
chore[#34]: SuccessCode 사용하도록 수정

sangyunpark99 reviewed (approved):
수고하셨습니다~

likerhythm committed:
refactor[#34]: User 엔티티 update문 추가

likerhythm committed:
feat[#34]: 여러 컨텐츠의 장르 한번에 찾는 쿼리 추가

likerhythm committed:
chore[#34]: return 위치 수정

likerhythm committed:
Merge branch 'develop' into feat/onboarding-submit

likerhythm committed:
Merge branch 'develop' into feat/onboarding-submit

likerhythm committed:
chore[#34]: 세미콜론 수정
